src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -237,6 +241,49 @@ void ObjectMonitor::operator delete[] (void *p) {
   operator delete(p);
 }
 
+// Check that object() and set_object() are called from the right context:
+static void check_object_context() {
+  Thread *self = Thread::current();
+  // ThreadService::get_current_contended_monitor() can call here via
+  // the VMThread so sanity check it.
+  guarantee(self->is_Java_thread() || self->is_VM_thread(), "must be");
+  if (self->is_Java_thread()) {
+    JavaThread* jt = (JavaThread*)self;
+    switch (jt->thread_state()) {
+    case _thread_in_vm:    // the usual case
+    case _thread_in_Java:  // during deopt
+      break;
+    default:
+      fatal("called from an unsafe thread state");
+    }
+    guarantee(jt->is_active_Java_thread(), "must be active JavaThread");
+  }
+}
+
+oop ObjectMonitor::object() const {
+  check_object_context();
+  if (_object.is_null()) {
+    return NULL;
+  }
+  return _object.resolve();
```

Comment:
Because before the TSM removal patch, the monitors are allocated in blocks. Then, the objects are not known when the monitors are allocated. The handles are instead lazily created... for now. With the next patch, that will probably change, if we agree to allocate ObjectMonitors one by one on inflation time. But that is one patch away.

Reply:
Erik's part3 patch ensures that when the ObjectMonitor is allocated,
the weak handle is created and initialized to the oop at that point. Since
this is still the TSM version, the ObjectMonitor may not yet have a weak
handle allocated at the time that object() is called.

Reply:
Ok, this is cool then and good that this logic is here instead of inside the WeakHandle::resolve/peek functions.  The OopHandle similar functions have a null check, so if this was permanent, I'd suggest putting it there.  Since this is going to change, having it here and keeping the non-null invariants for WeakHandle::resolve/peek functions is better.

Reply:
Thanks for confirmation.